### PR TITLE
🚨🚨 Source Monday migration to new api version 2024-01

### LINKED ISSUE #33579: Source Monday: connector migration to new 2024-01 API version
### Topic

Source Monday new API migration

### Relevant information

Support of current Monday's API version (2023-07) will end on [January 15th 2024](https://developer.monday.com/api-reference/docs/api-versioning#making-a-request).
Need to investigate and create a plan for connector migration to the new API version 2024-01.

### PR DESCRIPTION
**Pull request by midavadim**
## What
Fixed #33579 
migration to new api version 2024-01

## How
id fields migrated from int to str
removed obsolete fields
added new cursor pagination

## Recommended reading order
1. `x.java`
2. `y.python`

## 🚨 User Impact 🚨
YES,
id fields migrated from int to str

## Pre-merge Actions
*Expand the relevant checklist and delete the others.*

<details><summary><strong>New Connector</strong></summary>

### Community member or Airbyter

- **Community member?** Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- Connector version is set to `0.0.1`
    - `Dockerfile` has version `0.0.1`
- Documentation updated
    - Connector's `README.md`
    - Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - `docs/integrations/<source or destination>/<name>.md` including changelog with an entry for the initial version. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
    - `docs/integrations/README.md`

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- Create a non-forked branch based on this PR and test the below items on it
- Build is successful
- If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).

</details>

<details><summary><strong>Updating a connector</strong></summary>

### Community member or Airbyter

- Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- Unit & integration tests added

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- Create a non-forked branch based on this PR and test the below items on it
- Build is successful
- If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).

</details>

<details><summary><strong>Connector Generator</strong></summary>

- Issue acceptance criteria met
- PR name follows [PR naming conventions](https://docs.airbyte.com/contributing-to-airbyte/resources/pull-requests-handbook)
- If adding a new generator, add it to the [list of scaffold modules being tested](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connector-templates/generator/build.gradle#L41)
- The generator test modules (all connectors with `-scaffold` in their name) have been updated with the latest scaffold by running `./gradlew :airbyte-integrations:connector-templates:generator:generateScaffolds` then checking in your changes
- Documentation which references the generator is updated as needed

</details>

<details><summary><strong>Updating the Python CDK</strong></summary>

### Airbyter

Before merging:
- Pull Request description explains what problem it is solving
- Code change is unit tested
- Build and my-py check pass
- Smoke test the change on at least one affected connector
   - On Github: Run [this workflow](https://github.com/airbytehq/airbyte/actions/workflows/connectors_tests.yml), passing `--use-local-cdk --name=source-<connector>` as options
   - Locally: `airbyte-ci connectors --use-local-cdk --name=source-<connector> test`
- PR is reviewed and approved
      
After merging:
- [Publish the CDK](https://github.com/airbytehq/airbyte/actions/workflows/publish-cdk-command-manually.yml)
   - The CDK does not follow proper semantic versioning. Choose minor if this the change has significant user impact or is a breaking change. Choose patch otherwise.
   - Write a thoughtful changelog message so we know what was updated.
- Merge the platform PR that was auto-created for updating the Connector Builder's CDK version
   - This step is optional if the change does not affect the connector builder or declarative connectors.

</details>

### CODE REVIEW

**Review by tolik0 — APPROVED** [review]

File: airbyte-integrations/connectors/source-monday/integration_tests/sample_state.json
Comment (maxi297) [inline airbyte-integrations/connectors/source-monday/integration_tests/sample_state.json:9]: How is this change related to v2024-01?

File: airbyte-integrations/connectors/source-monday/unit_tests/test_graphql_requester.py
Comment (maxi297) [inline airbyte-integrations/connectors/source-monday/unit_tests/test_graphql_requester.py:157]: Why are `object_name`, `field_schema` and `stream_slice` defined? They don't seem to be used

File: airbyte-integrations/connectors/source-monday/unit_tests/test_graphql_requester.py
Comment (maxi297) [inline airbyte-integrations/connectors/source-monday/unit_tests/test_graphql_requester.py:56]: What is the use for this commented line?

File: airbyte-integrations/connectors/source-monday/unit_tests/test_extractor.py
Comment (maxi297) [inline airbyte-integrations/connectors/source-monday/unit_tests/test_extractor.py:54]: Is there a reason why not to provide `field_path_pagination` and `field_path_incremental` during `__init__`?

Furthermore, what does configuring `field_path_pagination` do for this test? It seems like it is unrelated for asserting `records == [{'id': 1}]`

File: airbyte-integrations/connectors/source-monday/unit_tests/test_extractor.py
Comment (maxi297) [inline airbyte-integrations/connectors/source-monday/unit_tests/test_extractor.py:56]: Why do we call `__post_init__` manually?

File: airbyte-integrations/connectors/source-monday/metadata.yaml
Comment (katmarkham) [inline airbyte-integrations/connectors/source-monday/metadata.yaml:17]: Source Monday has deprecated API version 2023-07.  We have upgraded the connector to the latest API version 2024-01.  In this new version, the Id field has changed from an integer to a string in the streams Boards, Items, Teams, Users, and Workspaces. Please reset affected streams.

File: docs/integrations/sources/monday-migrations.md
Comment (katmarkham) [inline docs/integrations/sources/monday-migrations.md:3]: please follow the migration guide template: https://docs.google.com/document/d/1V3ybA24Lj_fy7IQwsVVzSxs46qGOsUydwlAQEYSQfcE/edit?pli=1

**Review by katmarkham — CHANGES_REQUESTED** [review]

**Review by maxi297 — COMMENTED** [review]
Partial PR but I need to investigate more since I'm seeing discrepancy in behavior between the two versions (will communicate them synchronously if possible)

**Review by maxi297 — CHANGES_REQUESTED** [review]
I have identified the cause of the discrepancy I was seeing. Pagination seems broken for `activity_logs` at least. The reason is that we only do pagination on `boards` and not `activity_logs`. Before, requests like `https://api.monday.com/v2/?query=query{boards(limit:1,page:2){activity_logs(limit:50,page:3){id,event,data,entity,created_at}}}` were done. `page` in `activity_logs(limit:50,page:3)` is not sent in the version of the connector in this PR.

File: docs/integrations/sources/monday-migrations.md
Comment (midavadim) [reply docs/integrations/sources/monday-migrations.md:3]: Done

File: airbyte-integrations/connectors/source-monday/integration_tests/sample_state.json
Comment (midavadim) [reply airbyte-integrations/connectors/source-monday/integration_tests/sample_state.json]: removed

File: airbyte-integrations/connectors/source-monday/unit_tests/test_graphql_requester.py
Comment (midavadim) [reply airbyte-integrations/connectors/source-monday/unit_tests/test_graphql_requester.py:56]: removed

File: airbyte-integrations/connectors/source-monday/unit_tests/test_extractor.py
Comment (midavadim) [reply airbyte-integrations/connectors/source-monday/unit_tests/test_extractor.py:56]: it is supposed that field_path_pagination and  field_path_incremental come from manifest.yaml.But here we have a deal with 'empty' object

**Comment by midavadim** [comment]
> I have identified the cause of the discrepancy I was seeing. Pagination seems broken for `activity_logs` at least. The reason is that we only do pagination on `boards` and not `activity_logs`. Before, requests like `https://api.monday.com/v2/?query=query{boards(limit:1,page:2){activity_logs(limit:50,page:3){id,event,data,entity,created_at}}}` were done. `page` in `activity_logs(limit:50,page:3)` is not sent in the version of the connector in this PR.

It is fixed now. I tested on our Monday account and both connector version synced the same number of records but it is probably because we didn't need to use pagination.

**Comment by maxi297** [comment]
> It is fixed now. I tested on our Monday account and both connector version synced the same number of records but it is probably because we didn't need to use pagination.

I will not block this PR as it is urgent. However, can we add tests to validate pagination. It seems like a basic feature that we shouldn't need to test manually every time. I would suggest adding integration tests but if improving our dataset allows for that and that we don't expect much maintenance associated to this, this could be an ok solution too

File: airbyte-integrations/connectors/source-monday/unit_tests/test_graphql_requester.py
Comment (midavadim) [reply airbyte-integrations/connectors/source-monday/unit_tests/test_graphql_requester.py:157]: removed

File: airbyte-integrations/connectors/source-monday/unit_tests/test_extractor.py
Comment (midavadim) [reply airbyte-integrations/connectors/source-monday/unit_tests/test_extractor.py:54]: this particuar code tests extraction data from response during incremental sync.
Our code tries all possible paths: field_path, field_path_pagination and last field_path_incremental.

So here we test that previous paths failed and the last (field_path_pagination) succeded.

**Comment by midavadim** [comment]
> will not block this PR as it is urgent. However, can we add tests to validate pagination. It seems like a basic feature that we shouldn't need to test manually every time. I would suggest adding integration tests but if improving our dataset allows for that and that we don't expect much maintenance associated to this, this could be an ok solution too

Agree, I will add better tests based on  requests_mock. But it will take a significant amout of time. So I need your approval to do it after merge of current veersion.

File: airbyte-integrations/connectors/source-monday/source_monday/graphql_requester.py
Comment (maxi297) [inline airbyte-integrations/connectors/source-monday/source_monday/graphql_requester.py:108]: Is there a reason to pass `page=sub_page` here? As `sub_page==None`, I don't see the value added and it generates a lot of questions for the reader

File: airbyte-integrations/connectors/source-monday/source_monday/item_pagination_strategy.py
Comment (maxi297) [inline airbyte-integrations/connectors/source-monday/source_monday/item_pagination_strategy.py:88]: Can we clarify why we can just take the first one? My hypothesis based on [that](https://github.com/airbytehq/airbyte/blob/ea07171669381f355059720a11a2889c8f65f00b/airbyte-integrations/connectors/source-monday/source_monday/manifest.yaml#L182) is that we only request one board at the same time but it is very unclear for this context. Even the documentation is confusing as it states `This pagination strategy supports nested loop through 'boards' on the top level` but there is not really a capability to loop over multiple boards

**Review by maxi297 — APPROVED** [review]
I'll approve since this is urgent and the behavior seems fine. However, I would like to:
* Clean or document the code related to the comments in this round of review
* Add tests to ensure we don't inadvertently delete pagination logic that is needed

**Comment by maxi297** [comment]
> Agree, I will add better tests based on requests_mock

We have a framework to ease integration testing. I would suggest you sync with @askarpets when possible. If this framework does not work for your needs, please document why so that we can improve it

File: docs/integrations/sources/monday.md
Comment (katmarkham) [inline docs/integrations/sources/monday.md:77]: Migrated to the latest API version: 2024-01

**Review by katmarkham — CHANGES_REQUESTED** [review]

**Review by katmarkham — APPROVED** [review]

File: airbyte-integrations/connectors/source-monday/source_monday/item_pagination_strategy.py
Comment (midavadim) [reply airbyte-integrations/connectors/source-monday/source_monday/item_pagination_strategy.py:88]: there is always only one board due to limit=1, so in one request we extract all 'items_page' for one board only. 
We actually loop over boards: we 1 board per page.

File: airbyte-integrations/connectors/source-monday/source_monday/graphql_requester.py
Comment (midavadim) [reply airbyte-integrations/connectors/source-monday/source_monday/graphql_requester.py:108]: it is removed

**Pull request merged**
